ipq806x: add support for Arris TR4400 v2 / RAC2V1A

Lanchon opened the pull request:
note that this port requires moving the contents of a stock partition to consolidate usable flash space. the stock rom does not allow firmware upgrades and i did not find it vulnerable to known hacks, so the installation is non-trivial and requires serial access. thus, moving this partition data only requires a singled added command to a kind of long list of required actions and duplicates the available space for openwrt. the partition move can be easily rolled back to return to stock.

~i am in the process of documenting a detailed and safe install procedure which will be posted to the device page. the procedure includes a complete off-device backup of the flash.~

~also note that the device ends up with a main 155 MB `ubi` partition and an unused 64MB `extra` partition. i am including the `uvol` package in the build to ease usage of this extra partition if the user so desires, but i don't know if this inclusion will be frowned upon (the package could be installed manually if needed). i just want to provide an easy way for users to have an on-device partition that survives sysupgrades and wipes and is easey to remount afterwards.~

PS. i don't know how device additions are handled in the release cycle. i suppose it's way too late to backport to 22.x?

Lanchon commented:
stock flash layout:
```
[    1.803842] 0x000000000000-0x000000040000 : "0:SBL1"
[    1.810930] 0x000000040000-0x000000180000 : "0:MIBIB"
[    1.819310] 0x000000180000-0x0000002c0000 : "0:SBL2"
[    1.827320] 0x0000002c0000-0x000000540000 : "0:SBL3"
[    1.834944] 0x000000540000-0x000000660000 : "0:DDRCONFIG"
[    1.843184] 0x000000660000-0x000000780000 : "0:SSD"
[    1.851010] 0x000000780000-0x000000a00000 : "0:TZ"
[    1.862091] 0x000000a00000-0x000000c80000 : "0:RPM"
[    1.873377] 0x000000c80000-0x000001180000 : "0:APPSBL"                         5 MB
[    1.891176] 0x000001180000-0x000001200000 : "0:APPSBLENV"                    0.5 MB
[    1.897889] 0x000001200000-0x000001340000 : "0:ART"
[    1.906103] 0x000001340000-0x000005340000 : "rootfs"             64 MB
[    2.026199] 0x000005340000-0x0000053a0000 : "0:BOOTCONFIG"
[    2.033059] 0x0000053a0000-0x0000054e0000 : "0:SBL2_1"
[    2.041514] 0x0000054e0000-0x000005760000 : "0:SBL3_1"
[    2.049552] 0x000005760000-0x000005880000 : "0:DDRCONFIG_1"
[    2.056518] 0x000005880000-0x0000059a0000 : "0:SSD_1"
[    2.064463] 0x0000059a0000-0x000005c20000 : "0:TZ_1"
[    2.072265] 0x000005c20000-0x000005ea0000 : "0:RPM_1"
[    2.088589] 0x000005ea0000-0x000005f00000 : "0:BOOTCONFIG1"
[    2.095149] 0x000005f00000-0x000006400000 : "0:APPSBL_1"
[    2.112002] 0x000006400000-0x00000a400000 : "rootfs_1"           64 MB
[    2.251371] 0x00000a400000-0x00000a500000 : "fw_env"                           1 MB
[    2.258742] 0x00000a500000-0x00000ad00000 : "config"              8 MB
[    2.281909] 0x00000ad00000-0x00000af00000 : "PKI"                 2 MB
[    2.291273] 0x00000af00000-0x00000b000000 : "scfgmgr"             1 MB
               0x00000b000000-0x000010000000 -not used in stock-    80 MB

erase blocksize: 128 KB
```

new flash layout:
```
[    1.569462] 0x000000000000-0x000000040000 : "0:SBL1"
[    1.576043] 0x000000040000-0x000000180000 : "0:MIBIB"
[    1.582543] 0x000000180000-0x0000002c0000 : "0:SBL2"
[    1.587371] 0x0000002c0000-0x000000540000 : "0:SBL3"
[    1.598714] 0x000000540000-0x000000660000 : "0:DDRCONFIG"
[    1.601229] 0x000000660000-0x000000780000 : "0:SSD"
[    1.605872] 0x000000780000-0x000000a00000 : "0:TZ"
[    1.612956] 0x000000a00000-0x000000c80000 : "0:RPM"
[    1.617683] 0x000000c80000-0x000001180000 : "0:APPSBL"
[    1.626679] 0x000001180000-0x000001200000 : "0:APPSBLENV"
[    1.628128] 0x000001200000-0x000001340000 : "0:ART"
D[    1.633716] 0x000001340000-0x000005340000 : "stock_rootfs"
[    1.742708] 0x000005340000-0x0000053a0000 : "0:BOOTCONFIG"
[    1.744431] 0x0000053a0000-0x0000054e0000 : "0:SBL2_1"
[    1.750457] 0x0000054e0000-0x000005760000 : "0:SBL3_1"
[    1.757753] 0x000005760000-0x000005880000 : "0:DDRCONFIG_1"
[    1.760882] 0x000005880000-0x0000059a0000 : "0:SSD_1"
[    1.765758] 0x0000059a0000-0x000005c20000 : "0:TZ_1"
[    1.773655] 0x000005c20000-0x000005ea0000 : "0:RPM_1"
[    1.779160] 0x000005ea0000-0x000005f00000 : "0:BOOTCONFIG1"
[    1.780916] 0x000005f00000-0x000006400000 : "0:APPSBL_1"
D[    1.793839] 0x000006400000-0x00000a400000 : "stock_rootfs_1"
D[    1.913928] 0x00000a400000-0x00000a500000 : "stock_fw_env"
D[    1.916203] 0x00000a500000-0x00000ad00000 : "stock_config"
D[    1.932297] 0x00000ad00000-0x00000af00000 : "stock_PKI"
D[    1.936177] 0x00000af00000-0x00000b000000 : "stock_scfgmgr"
N[    1.938529] 0x000006400000-0x000006500000 : "fw_env"                          1 MB
N[    1.944230] 0x000006500000-0x000010000000 : "ubi"              155 MB
N[    2.205033] 0x000001340000-0x000005340000 : "extra"             64 MB

D: deprecated partitions. kept to support rollback. prefixed by "stock_". all are R/W.
N: new partitions. "fw_env" is read-only.
```

dangowrt commented on target/linux/ipq806x/image/generic.mk:
```diff
@@ -65,6 +65,19 @@ define Device/ZyXELImage
+	SOC := qcom-ipq8065
+	BLOCKSIZE := 128k
+	PAGESIZE := 2048
+	DEVICE_PACKAGES := ath10k-firmware-qca9984-ct ath10k-firmware-qca99x0-ct uvol
```
While it's cool to see you want to use `uvol`, for now uvol is only available in the packages feed and can hence not be added as a default package.

Lanchon replied:
uvol removed, thanks!

dangowrt reviewed (changes requested):
If this is supposed to get merged to openwrt.git you will need to use your full real name in the
`Signed-off-by:` line. Once this is more complete (ie. correct SoB, instructions for flashing put into commit message, ...) we can backport it to 22.03.0.

Lanchon commented:
thanks! uvol removed. SoB fixed. instructions follow hopefully tomorrow.

mans0n commented on target/linux/ipq806x/base-files/etc/board.d/02_network:
```diff
@@ -11,6 +11,11 @@ board_config_update
 board=$(board_name)
 
 case "$board" in
+arris,tr4400-v2)
+	ucidef_set_interfaces_lan_wan "eth1" "eth2"
+	ucidef_add_switch "switch0" \
+		"1:lan" "2:lan" "3:lan" "4:lan" "6u@eth1" "0u@eth0"
+	;;
```
Is the switch connected to eth0? Could you explain how eth1 and eth2 are used for this device?

Lanchon replied:
yes, both eth0 and eth1 are (seem to be) connected to the switch in ports 0 and 6. both can be used for lan. i tested both and chose eth1 for lan because of prior art (a v1 port -not upstreamed- chose that) and because eth1 seems to be commonly used for lan in the devices i've seen.

now here is a quirk: the wan port seems not to be connected to the switch and seems connected to eth2 directly instead. `ip a` shows up/down status according to whether a cable is connected or not to the wan port. however luci shows eth2 as "connected: yes" regardless of actual connection.

besides this detail, the 5 ports work as expected.

Lanchon replied:
btw, in the stock fw: eth0 is unused, eth1 is lan, and eth2 is wan. (same in this port.)

mans0n replied:
Okay, that explains. Thanks for the info. Have no idea for the luci.

mans0n commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-tr4400-v2.dts:
```diff
@@ -0,0 +1,375 @@
+
+		flash@0 {
+			compatible = "everspin,mr25h256";
+			#address-cells = <1>;
+			#size-cells = <1>;
```
IMO these two lines can be dropped safely.

Lanchon replied:
removed two instances of these lines.

mans0n commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-tr4400-v2.dts:
```diff
@@ -0,0 +1,375 @@
+&gmac0 {
+	status = "okay";
+	phy-mode = "rgmii";
+	qcom,id = <0>;
```
IIRC `qcom,id` can be replaced by `ethernetX`.
https://github.com/openwrt/openwrt/blob/72b93b77a5b406d0a32740d3a000218e19451460/target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-g10.dts#L12

Lanchon replied:
i'm sorry, could you elaborate? i don't understand the reason for this change and i also cannot find relevant alias definitions. thanks!

mans0n replied:
Sorry, I was mistaken, please ignore this.

mans0n commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-tr4400-v2.dts:
```diff
@@ -0,0 +1,375 @@
+	pinctrl-0 = <&rgmii2_pins>;
+	pinctrl-names = "default";
+
+	mtd-mac-address = <&fw_env 0x18>;
```
`mtd-mac-address` should not work since 5ae2e786395c, you should migrate to `nvmem-cells` instead. (While at it you might also try `nvmem-cells` for ath10k precal)
https://github.com/openwrt/openwrt/blob/72b93b77a5b406d0a32740d3a000218e19451460/target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts#L254
https://github.com/openwrt/openwrt/blob/72b93b77a5b406d0a32740d3a000218e19451460/target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8064-r7500v2.dts#L200

Lanchon replied:
> `mtd-mac-address` should not work since [5ae2e78](https://github.com/openwrt/openwrt/commit/5ae2e786395c7f9db0167ebe875be5df9502d8d8), you should migrate to `nvmem-cells` instead.

you are right! ethX macs were not being set, now they are. thanks!
the fifth mac addr in fw_env i haven't seen used in stock fw. also i never saw eth0 enabled and its mac programmed. i decided to use it for eth0, but i don't know if stock did the same thing or not.

Question: (im very new to OpenWrt) wlan0 and 1 differ in mac address by 1 value. is that ok if virtual APs are configured? in DD-WRT that would cause collisions.

Lanchon replied:
> While at it you might also try `nvmem-cells` for ath10k precal

i think these wlan adapters are enumerated over pcie instead of configured in the device tree, am i not right? i don't know how i'd go about about configuring them via the dts. right now the wlan macs are being correctly configured

mans0n commented:
And could you copy the detailed device specification in the wiki page to the commit message?

Lanchon commented:
@mans0n 

> could you copy the detailed device specification in the wiki page to the commit message?

the problem with the spec (i wrote the wiki) is, i'm not sure of it. the FCC ID printed in the case seems to be for a different device (v1?). that FCC ID corresponds to different hardware (both ram and flash size are different, GPIO usage seems to be different), but i guess the wifi module should be the same and thus the shared FCC ID. i guess the soc is IPQ8065 but i'm not even sure of that. all chips are covered with heatsinks and RF shields and i'm not willing to tear that apart. to err on the side of caution, i tried to be scant and leave out all guess work from the commit, believing missing info is better than wrong info. the wiki can be amended more easily.

so i guess my question is: what information do you want added? if i add it all, then some of it might be incorrect.

thanks!


PS. i'll take a look at nvmem-cells ASAP.

Lanchon commented:
@mans0n and since you are here i might as well ask you about another detail that bugs me.

in the FCC ID entry they tear down the wifi pcie card and find a 4x4 QCA9984 for 5 GHz and a 3x3 QCA9983 for 2.4 GHz. my device is different, but i suppose it has the same wifi card, though i haven't lifted the heatsinks. i can see that i have 3 "large" antennas connected to both chips and a fourth smaller antenna connected to only one chip. the larger ones should be dual band, so the wifi card at least seems to be 3x3 2.4GHz and 4x4 5GHz as expected. so the working hypothesis is: same card as the FCC ID.

now, luci shows the 5 GHz adapter correctly:
Type: Qualcomm Atheros QCA9984 802.11nac

but the 2.4 GHz adapter is shown (incorrectly?) as:
Type: Qualcomm Atheros QCA9990 802.11bgn

am i missing something here? both adapters work well in my tests.

thanks!

mans0n commented:
> that FCC ID corresponds to different hardware (both ram and flash size are different, GPIO usage seems to be different)

What made you think so? I've skimmed over the wiki page and the fccid page but the pictures looked mostly same. The different RAM and flash sizes do not always mean the different board version. Different GPIO usage is a slightly bigger issue, could you explain more about it?

> so i guess my question is: what information do you want added? if i add it all, then some of it might be incorrect.

It seems the only unknown parts which are burried under the shields are SoC, RAM, and Wi-Fi chips, right? I think you can add info about the rest of them. (like Ethernet or USB port) And the flashing instruction should be added too. (refer to: 70c12d26ca6e da8428d277cd)

> but the 2.4 GHz adapter is shown (incorrectly?) as:
> Type: Qualcomm Atheros QCA9990 802.11bgn

Hmm, could you tell me the result of `lspci -nn`?

Lanchon commented:
@mans0n thanks!

> I think you can add info about the rest of them. (like Ethernet or USB port)

 ok, i will.

> And the flashing instruction should be added too.

i'm working on it. will require some guidance when ready, i will ping you.

> could you tell me the result of lspci -nn?

lspci is not available in stock nor in my port. what pkg do i install?

mans0n commented:
> lspci is not available in stock nor in my port. what pkg do i install?

`lspci` is available in `pciutils` package.

Lanchon commented:
@mans0n

> > that FCC ID corresponds to different hardware (both ram and flash size are different, GPIO usage seems to be different)
> 
> What made you think so? I've skimmed over the wiki page and the fccid page but the pictures looked mostly same. The different RAM and flash sizes do not always mean the different board version. Different GPIO usage is a slightly bigger issue, could you explain more about it?

check this out (supposedly a "TR4400-AC" (v1?) - mine had tr4400v2 all over the stock fw):
https://github.com/coolsnowwolf/lede/blob/f27a1742d6b7d4e48b481d8aae5fb9f0da5f497a/target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-rac2v1a.dts#L32-L41

compare to my device:
https://github.com/Lanchon/openwrt/blob/9da5c676f4a1cffafda4b333813f708c2275c383/target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-tr4400-v2.dts#L32-L41

i can't vouch for the correctness of the other port, but mine is correct. and i assume the other is right too (that dts seems to be in part decompiled form that router stock dtb). so i'm not stating that it is a different board, but there is compelling evidence that it might be (starting with the name).

take away: i'm not sure about the hw... except for the wifi card which should be the same given the same FCC ID.

Lanchon commented:
@mans0n 
```
root@OpenWrt:/# lspci -nn
0000:00:00.0 PCI bridge [0604]: Qualcomm Device [17cb:0101]
0000:01:00.0 Network controller [0280]: Qualcomm Atheros QCA9984 802.11ac Wave 2 Wireless Network Adapter [168c:0046]
0001:00:00.0 PCI bridge [0604]: Qualcomm Device [17cb:0101]
0001:01:00.0 Network controller [0280]: Qualcomm Atheros QCA9980/9990 802.11ac Wireless Network Adapter [168c:0040]
```

Lanchon commented:
my original question: https://github.com/openwrt/openwrt/pull/9623#issuecomment-1086634478

Lanchon commented:
@dangowrt, @mans0n,

so the flashing instructions... the device requires serial access to flash, and though it is possible to install via uboot, it is easier and safer to fully boot the stock fw and work there. (pressing enter will drop you to a root shell.)

it'd be great to download the required files from the console. but no curl, and bb wget does not support HTTPS. i could download via HTTP and then check hash, but unfortunately neither github nor gitlab serve regular HTTP.

so next best thing is download to a PC and push from there.

the installation is basically:
1. use the serial console to enable ssh.
2. use ssh to verify that the router state is exactly as expected.
3. use scp to upload initramfs image and install script.
4. run install script.

detail instructions: https://github.com/Lanchon/openwrt-tr4400-v2
the install script itself: https://github.com/Lanchon/openwrt-tr4400-v2/blob/main/install-recovery.sh

so how do i go about fitting this into the commit msg?
- a link to the instructions?
- do you want to clone the instructions and link to your clone?
- link to the device page and from there to instructions?
- something else?

thanks for the help!

(still pending: nvmem-cells stuff)

mans0n commented:
> check this out (supposedly a "TR4400-AC" (v1?) - mine had tr4400v2 all over the stock fw):
> https://github.com/coolsnowwolf/lede/blob/f27a1742d6b7d4e48b481d8aae5fb9f0da5f497a/target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-rac2v1a.dts#L32-L41

Thanks for the pointer. https://github.com/coolsnowwolf/lede/pull/7680
Does the stock FW show the "tr4400v2" string? I couldn't find it in the boot log from the wiki page.
@AmadeusGhost Do you have an idea for this?

For the QCA9983, I'm not sure but it seems it's normal. EA7500 has QCA9982 and QCA9983 but both of them have the id of 168c:0040.
https://openwrt.org/toh/linksys/linksys_ea7500_v1

AmadeusGhost commented:
> Does the stock FW show the "tr4400v2" string? I couldn't find it in the boot log from the wiki page. Do you have an idea for this?

I'm sorry I don't remember this, maybe 'Arris TR4400-AC' would be more appropriate, unless we know what is the difference between v1 and v2.
> i can't vouch for the correctness of the other port, but mine is correct.

Actually it was my mistake, the gpio of the button I copied from the decompiled dts, and the reset button doesn't work.
```
		gpio_keys {
			compatible = "gpio-keys";
			pinctrl-0 = < 0x27 >;
			pinctrl-names = "default";

			button@1 {
				label = "reset";
				linux,code = < 0x198 >;
				gpios = < 0x1a 0x36 0x01 >;
				linux,input-type = < 0x01 >;
				debounce-interval = < 0x3c >;
			};

			button@2 {
				label = "wps";
				linux,code = < 0x211 >;
				gpios = < 0x1a 0x41 0x01 >;
				linux,input-type = < 0x01 >;
				debounce-interval = < 0x3c >;
			};
		};
```

Lanchon commented:
@mans0n 

> Does the stock FW show the "tr4400v2" string? I couldn't find it in the boot log from the wiki page.

yep, looks very much like mine is a v2:
```
rod@rod-latitude:~/hardware/routers/arris-tr4400-v2/backups/expanded_rootfs$ grep -siIr 'tr4400ac'

rod@rod-latitude:~/hardware/routers/arris-tr4400-v2/backups/expanded_rootfs$ grep -siIr 'tr4400.ac'

rod@rod-latitude:~/hardware/routers/arris-tr4400-v2/backups/expanded_rootfs$ grep -siIr 'tr4400.*ac'

rod@rod-latitude:~/hardware/routers/arris-tr4400-v2/backups/expanded_rootfs$ grep -siIr 'tr4400v2'
etc/samba/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/apps/samba_3200/etc/samba
etc/samba/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
etc/samba.conf/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/apps/samba_3200/etc/samba.conf
etc/samba.conf/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
etc/svn.info:URL: svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source
etc/svn.info:Repository Root: svn://10.10.67.238:36901/svn/Router/TR4400v2
etc/svn.info:Build from: sercomm-QCA-Server /home2/mark/TR4400v2_0419b_0301/Source by mark
usr/lib/libstdc++.so.6.0.19-gdb.py:pythondir = '/home2/mark/TR4400v2_0419b_0301/Kernel/3_0_1500_216/qsdk/staging_dir/toolchain-arm_cortex-a7_gcc-4.8-linaro_uClibc-0.9.33.2_eabi/share/gcc-4.8.3/python'
usr/lib/libstdc++.so.6.0.19-gdb.py:libdir = '/home2/mark/TR4400v2_0419b_0301/Kernel/3_0_1500_216/qsdk/staging_dir/toolchain-arm_cortex-a7_gcc-4.8-linaro_uClibc-0.9.33.2_eabi/arm-openwrt-linux-uclibcgnueabi/lib'
usr/share/cups/mime/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/apps/cups/cups-1.4.6/conf
usr/share/cups/mime/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
www/Fonts/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/UI/gui.arris.spt/Fonts
www/Fonts/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
www/css/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/UI/gui.arris.spt/css
www/css/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
www/css/base/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/UI/gui.arris.spt/css/base
www/css/base/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
www/css/base/images/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/UI/gui.arris.spt/css/base/images
www/css/base/images/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
www/css/ui-lightness/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/UI/gui.arris.spt/css/ui-lightness
www/css/ui-lightness/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
www/css/ui-lightness/images/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/UI/gui.arris.spt/css/ui-lightness/images
www/css/ui-lightness/images/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
www/i/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/UI/gui.arris.spt/i
www/i/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
www/i/unused/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/UI/gui.arris.spt/i/unused
www/i/unused/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
www/lang/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/UI/gui.arris.spt/lang
www/lang/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
www/scripts/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/UI/gui.arris.spt/scripts
www/scripts/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
www/scripts/plugins/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/UI/gui.arris.spt/scripts/plugins
www/scripts/plugins/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2

rod@rod-latitude:~/hardware/routers/arris-tr4400-v2/backups/expanded_rootfs$ grep -siIr 'tr4400'
etc/inittab:::sysinit:/etc/init.d/rcS tr4400_sc_gwup boot
etc/samba/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/apps/samba_3200/etc/samba
etc/samba/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
etc/samba.conf/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/apps/samba_3200/etc/samba.conf
etc/samba.conf/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
etc/svn.info:URL: svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source
etc/svn.info:Repository Root: svn://10.10.67.238:36901/svn/Router/TR4400v2
etc/svn.info:Build from: sercomm-QCA-Server /home2/mark/TR4400v2_0419b_0301/Source by mark
etc/svn.info:BOARD_ID: TR4400-CH
usr/lib/libstdc++.so.6.0.19-gdb.py:pythondir = '/home2/mark/TR4400v2_0419b_0301/Kernel/3_0_1500_216/qsdk/staging_dir/toolchain-arm_cortex-a7_gcc-4.8-linaro_uClibc-0.9.33.2_eabi/share/gcc-4.8.3/python'
usr/lib/libstdc++.so.6.0.19-gdb.py:libdir = '/home2/mark/TR4400v2_0419b_0301/Kernel/3_0_1500_216/qsdk/staging_dir/toolchain-arm_cortex-a7_gcc-4.8-linaro_uClibc-0.9.33.2_eabi/arm-openwrt-linux-uclibcgnueabi/lib'
usr/share/cups/mime/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/apps/cups/cups-1.4.6/conf
usr/share/cups/mime/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
www/Fonts/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/UI/gui.arris.spt/Fonts
www/Fonts/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
www/IntConWiz_01.html:<title>ARRIS TR4400</title>
www/IntConWiz_02.html:<title>ARRIS TR4400</title>
www/IntConWiz_03.html:<title>ARRIS TR4400</title>
www/IntConWiz_04.html:<title>ARRIS TR4400</title>
www/IntConWiz_05.html:<title>ARRIS TR4400</title>
www/IntConWiz_06.html:<title>ARRIS TR4400</title>
www/RgFiltering.html:<title>ARRIS TR4400</title>
www/SRT_new.html:<title>ARRIS TR4400</title>
www/add_static_client_ip.html:<title>ARRIS TR4400</title>
www/b_device.html:<title>ARRIS TR4400</title>
www/basic_sett.html:<title>ARRIS TR4400</title>
www/chkpbc.html:<title>ARRIS TR4400</title>
www/chkpin.html:<title>ARRIS TR4400</title>
www/css/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/UI/gui.arris.spt/css
www/css/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
www/css/base/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/UI/gui.arris.spt/css/base
www/css/base/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
www/css/base/images/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/UI/gui.arris.spt/css/base/images
www/css/base/images/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
www/css/ui-lightness/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/UI/gui.arris.spt/css/ui-lightness
www/css/ui-lightness/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
www/css/ui-lightness/images/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/UI/gui.arris.spt/css/ui-lightness/images
www/css/ui-lightness/images/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
www/ddns.html:<title>ARRIS TR4400</title>
www/errpin1.html:<title>ARRIS TR4400</title>
www/errpin2.html:<title>ARRIS TR4400</title>
www/fw6_filter.html:<title>ARRIS TR4400</title>
www/fw6_filter_edit.html:<title>ARRIS TR4400</title>
www/fw6_filter_new.html:<title>ARRIS TR4400</title>
www/fw_alg.html:<title>ARRIS TR4400</title>
www/fw_dmz.html:<title>ARRIS TR4400</title>
www/fw_macfilter.html:<title>ARRIS TR4400</title>
www/fw_macfilter_new.html:<title>ARRIS TR4400</title>
www/fw_main.html:<title>ARRIS TR4400</title>
www/fw_ping.html:<title>ARRIS TR4400</title>
www/fw_pt.html:<title>ARRIS TR4400</title>
www/fw_pt_edit.html:<title>ARRIS TR4400</title>
www/fw_pt_new.html:<title>ARRIS TR4400</title>
www/fw_remote.html:<title>ARRIS TR4400</title>
www/fw_virt.html:<title>ARRIS TR4400</title>
www/i/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/UI/gui.arris.spt/i
www/i/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
www/i/unused/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/UI/gui.arris.spt/i/unused
www/i/unused/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
www/index.html:<title>ARRIS TR4400</title>
www/ip6_lan_sett.html:<title>ARRIS TR4400</title>
www/ip6_mapt.html:<title>ARRIS TR4400</title>
www/ip6_wan_dyna.html:<title>ARRIS TR4400</title>
www/ip6_wan_stat.html:<title>ARRIS TR4400</title>
www/lan_dhcp.html:<title>ARRIS TR4400</title>
www/lan_sett.html:<title>ARRIS TR4400</title>
www/lang/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/UI/gui.arris.spt/lang
www/lang/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
www/login.html:<title>ARRIS TR4400</title>
www/logout.html:<title>ARRIS TR4400</title>
www/mapt_useful_port.html:<title>ARRIS TR4400</title>
www/port_status.html:<title>ARRIS TR4400</title>
www/qos.html:<title>ARRIS TR4400</title>
www/reset_ok.html:<title>ARRIS TR4400</title>
www/resto_ok.html:<title>ARRIS TR4400</title>
www/rip.html:<title>ARRIS TR4400</title>
www/rip_1to1_setup.html:<title>ARRIS TR4400</title>
www/rip_tables_edit.html:<title>ARRIS TR4400</title>
www/ripng.html:<title>ARRIS TR4400</title>
www/sb_download.html:<title>ARRIS TR4400</title>
www/sb_priority.html:<title>ARRIS TR4400</title>
www/sb_sett.html:<title>ARRIS TR4400</title>
www/sb_time.html:<title>ARRIS TR4400</title>
www/scripts/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/UI/gui.arris.spt/scripts
www/scripts/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
www/scripts/plugins/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2/TR4400v2_0419b/Source/UI/gui.arris.spt/scripts/plugins
www/scripts/plugins/.svn/entries:svn://10.10.67.238:36901/svn/Router/TR4400v2
www/srt_setup.html:<title>ARRIS TR4400</title>
www/ssh.html:<title>ARRIS TR4400</title>
www/status.html:<title>ARRIS TR4400</title>
www/tr069.html:<title>ARRIS TR4400</title>
www/tr069_ca_acs.html:<title>ARRIS TR4400</title>
www/tr069_ca_dls.html:<title>ARRIS TR4400</title>
www/upg_ok.html:<title>ARRIS TR4400</title>
www/usb_acl.html:<title>ARRIS TR4400</title>
www/usb_edit_fld.html:<title>ARRIS TR4400</title>
www/usb_file.html:<title>ARRIS TR4400</title>
www/usb_fld_tree.html:<title>ARRIS TR4400</title>
www/usb_media.html:<title>ARRIS TR4400</title>
www/usb_new_fld.html:<title>ARRIS TR4400</title>
www/usb_status.html:<title>ARRIS TR4400</title>
www/ut_defau.html:<title>ARRIS TR4400</title>
www/ut_facty.html:<title>ARRIS TR4400</title>
www/ut_firew.html:<title>ARRIS TR4400</title>
www/ut_log.html:<title>ARRIS TR4400</title>
www/ut_prev.html:<title>ARRIS TR4400</title>
www/ut_reset.html:<title>ARRIS TR4400</title>
www/ut_save.html:<title>ARRIS TR4400</title>
www/ut_sys.html:<title>ARRIS TR4400</title>
www/wan_dns.html:<title>ARRIS TR4400</title>
www/wan_dyna.html:<title>ARRIS TR4400</title>
www/wan_pppoe.html:<title>ARRIS TR4400</title>
www/wan_pptp.html:<title>ARRIS TR4400</title>
www/wan_stat.html:<title>ARRIS TR4400</title>
www/wls_adv.html:<title>ARRIS TR4400</title>
www/wls_atf.html:<title>ARRIS TR4400</title>
www/wls_chan.html:<title>ARRIS TR4400</title>
www/wls_gst.html:<title>ARRIS TR4400</title>
www/wls_log.html:<title>ARRIS TR4400</title>
www/wls_qos_classification.html:<title>ARRIS TR4400</title>
www/wls_status.html:<title>ARRIS TR4400</title>
www/wls_wps.html:<title>ARRIS TR4400</title>
```
do you want the userland?

> For the QCA9983, I'm not sure but it seems it's normal.

cool, i'll let that be then, thanks.

@AmadeusGhost hey! thanks for your work!

where did you see the string `Arris TR4400-AC`? it's not in my fw at all AFAICT.

Lanchon commented:
@dangowrt @mans0n ,

i think i've addressed all your comments, except:
- install instructions: [please comment](https://github.com/openwrt/openwrt/pull/9623#issuecomment-1086827902) on how to proceed.
- nvmem-cells for ath10k precal: i've looked at this. i think it should be a separate commit, as i want this device backported to 22.x. later i can give it a try.

thanks!

mans0n commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-tr4400-v2.dts:
```diff
@@ -0,0 +1,387 @@
+		};
+	};
+
+	rgmii2_pins: rgmii2_pins {
+		tx {
+			pins = "gpio27", "gpio28", "gpio29", "gpio30", "gpio31", "gpio32";
+			input-disable;
+		};
+	};
```
What happens if you remove this node? `&rgmii2_pins` node is already defined in the dtsi so it might be possible that you don't need these extra definition.

Lanchon replied:
i don't know much of dev trees, just encountered them for the first time now.

**all** ipq8085 ports have these lines, so my port has them too.

the included lines are mux, not tx (the significance to me is unknown):

```
   rgmii2_pins: rgmii2_pins {
    mux {
     pins = "gpio27", "gpio28", "gpio29",
            "gpio30", "gpio31", "gpio32",
            "gpio51", "gpio52", "gpio59",
            "gpio60", "gpio61", "gpio62";
     function = "rgmii2";
     drive-strength = <8>;
     bias-disable;
    };
   };
```

but note that RT4230W does seem to have duplicate mux entries, that one probably needs fixing.

mans0n replied:
I did the pinctrl refactoring in 19c8f11d92f3 but I forgot that ipq8065 devices had the tx node even at that time, sorry again.

mans0n commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-tr4400-v2.dts:
```diff
@@ -0,0 +1,387 @@
+	pinctrl-0 = <&mdio0_pins>;
+	pinctrl-names = "default";
+
+	phy0: ethernet-phy@0 {
```
I think you can drop this label.

Lanchon replied:
i don't follow. what do you mean? which lines?

mans0n replied:
I mean, `phy0` label is not referenced anywhere.
```diff
-	phy0: ethernet-phy@0 {
+	ethernet-phy@0 {
```

Lanchon replied:
thanks!

mans0n commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-tr4400-v2.dts:
```diff
@@ -0,0 +1,387 @@
+			>;
+	};
+
+	phy4: ethernet-phy@4 {
+		reg = <4>;
+	};
```
What happens if you drop this node and remove `phy-handle` in `&gmac0`?

Lanchon replied:
@mans0n,

i supposed the same applies to phy7. i removed both: eth0 (phy4) worked, but eth2 did not.

this is a total guess game for me. i have to tell you i've more or less exhausted the time i was willing to spend into upstreaming this. also this is the first time i even looked at a dts. i'm definitely not the right man to make improvements, i just mixed and matched from other ports and the stock decompiled dtb. in fact, i've never even used openwrt or had an openwrt router before.

my priority here is backporting to 22.x when that happens and hopefully have an unofficial build that can use the official repos. i also want to contribute to other areas of openwrt, such as a safer sysupgrade mechanism; hopefully i'll have the time.

mans0n replied:
> i supposed the same applies to phy7. i removed both: eth0 (phy4) worked, but eth2 did not.

Sorry, I should have provided the context.
Actually that matches my expectation. eth0 and eth1 are connected to the switch and the ar8327 swconfig driver seems to not require phy-handle. But eth2 is connected to AR8033 phy and that needs phy-handle so it cannot be skipped.

mans0n commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-tr4400-v2.dts:
```diff
@@ -0,0 +1,387 @@
+	status = "okay";
+};
+
+&usb3_0 {
+	status = "okay";
+	clocks = <&gcc USB30_1_MASTER_CLK>;
+};
+
+&usb3_1 {
+	status = "okay";
+	clocks = <&gcc USB30_0_MASTER_CLK>;
+};
```
Your device has only one USB port, so I guess you need only one node?

Lanchon replied:
yes, this is true, and more serious. IDK if i have the time though. note that the stock dtb has these two entries.

Lanchon replied:
```
root@OpenWrt:/# lsusb 
Bus 002 Device 001: ID 1d6b:0003 Linux 5.10.109 xhci-hcd xHCI Host Controller
Bus 004 Device 001: ID 1d6b:0003 Linux 5.10.109 xhci-hcd xHCI Host Controller
Bus 001 Device 002: ID 0a12:0001  BT DONGLE10
Bus 001 Device 001: ID 1d6b:0002 Linux 5.10.109 xhci-hcd xHCI Host Controller
Bus 003 Device 001: ID 1d6b:0002 Linux 5.10.109 xhci-hcd xHCI Host Controller
root@OpenWrt:/# lspci 
0000:00:00.0 PCI bridge: Qualcomm Device 0101
0000:01:00.0 Network controller: Qualcomm Atheros QCA9984 802.11ac Wave 2 Wireless Network Adapter
0001:00:00.0 PCI bridge: Qualcomm Device 0101
0001:01:00.0 Network controller: Qualcomm Atheros QCA9980/9990 802.11ac Wireless Network Adapter
```

@mans0n no way that this could be a usb (minus some kind of phy)?
https://openwrt.org/inbox/toh/arris/tr4400_v2?do=#jtag

Lanchon replied:
seems to have 2 hci pairs:
```
root@OpenWrt:/# dmesg | egrep -i '(usb|(o|u|e|x)hci)'
[    0.066633] usbcore: registered new interface driver usbfs
[    0.066721] usbcore: registered new interface driver hub
[    0.066793] usbcore: registered new device driver usb
[    6.860938] ehci_hcd: USB 2.0 'Enhanced' Host Controller (EHCI) Driver
[    6.882766] ehci-fsl: Freescale EHCI Host controller driver
[    6.883447] ehci-platform: EHCI generic platform driver
[    6.900598] ohci_hcd: USB 1.1 'Open' Host Controller (OHCI) Driver
[    6.901087] ohci-platform: OHCI generic platform driver
[    6.907661] dwc3-qcom 110f8800.usb3: IRQ hs_phy_irq not found
[    6.910867] dwc3-qcom 110f8800.usb3: IRQ dp_hs_phy_irq not found
[    6.916689] dwc3-qcom 110f8800.usb3: IRQ dm_hs_phy_irq not found
[    6.922852] dwc3-qcom 110f8800.usb3: IRQ ss_phy_irq not found
[    6.930647] dwc3-qcom 100f8800.usb3: IRQ hs_phy_irq not found
[    6.934397] dwc3-qcom 100f8800.usb3: IRQ dp_hs_phy_irq not found
[    6.940201] dwc3-qcom 100f8800.usb3: IRQ dm_hs_phy_irq not found
[    6.946203] dwc3-qcom 100f8800.usb3: IRQ ss_phy_irq not found
[    7.093340] xhci-hcd xhci-hcd.0.auto: xHCI Host Controller
[    7.093389] xhci-hcd xhci-hcd.0.auto: new USB bus registered, assigned bus number 1
[    7.097876] xhci-hcd xhci-hcd.0.auto: hcc params 0x0228f065 hci version 0x100 quirks 0x0000000002010010
[    7.105400] xhci-hcd xhci-hcd.0.auto: irq 51, io mem 0x11000000
[    7.115602] hub 1-0:1.0: USB hub found
[    7.124562] xhci-hcd xhci-hcd.0.auto: xHCI Host Controller
[    7.128452] xhci-hcd xhci-hcd.0.auto: new USB bus registered, assigned bus number 2
[    7.133824] xhci-hcd xhci-hcd.0.auto: Host supports USB 3.0 SuperSpeed
[    7.148011] usb usb2: We don't know the algorithms for LPM for this host, disabling LPM.
[    7.148351] hub 2-0:1.0: USB hub found
[    7.160057] xhci-hcd xhci-hcd.1.auto: xHCI Host Controller
[    7.163784] xhci-hcd xhci-hcd.1.auto: new USB bus registered, assigned bus number 3
[    7.169221] xhci-hcd xhci-hcd.1.auto: hcc params 0x0228f065 hci version 0x100 quirks 0x0000000002010010
[    7.176793] xhci-hcd xhci-hcd.1.auto: irq 52, io mem 0x10000000
[    7.187249] hub 3-0:1.0: USB hub found
[    7.196048] xhci-hcd xhci-hcd.1.auto: xHCI Host Controller
[    7.199902] xhci-hcd xhci-hcd.1.auto: new USB bus registered, assigned bus number 4
[    7.205113] xhci-hcd xhci-hcd.1.auto: Host supports USB 3.0 SuperSpeed
[    7.212869] usb usb4: We don't know the algorithms for LPM for this host, disabling LPM.
[    7.220016] hub 4-0:1.0: USB hub found
[    7.569534] usb 1-1: new full-speed USB device number 2 using xhci-hcd
```

but there is something strange with the clock defs. they are swapped?
```
  usb3_0: usb3@110f8800 {
   compatible = "qcom,dwc3", "syscon";
   #address-cells = <1>;
   #size-cells = <1>;
   reg = <0x110f8800 0x8000>;
   clocks = <&gcc 267>;
   clock-names = "core";
...

&usb3_0 {
 status = "okay";
 clocks = <&gcc 268>;
};
```
```
  usb3_1: usb3@100f8800 {
   compatible = "qcom,dwc3", "syscon";
   #address-cells = <1>;
   #size-cells = <1>;
   reg = <0x100f8800 0x8000>;
   clocks = <&gcc 268>;
   clock-names = "core";
...

&usb3_1 {
 status = "okay";
 clocks = <&gcc 267>;
};
```

i'll try leaving clock defs out.

Lanchon replied:
works without the clock defs.

Lanchon replied:
these drivers are loaded but seem unused, as the hw is xhci.
```
[    7.450651] ehci_hcd: USB 2.0 'Enhanced' Host Controller (EHCI) Driver
[    7.471760] ehci-fsl: Freescale EHCI Host controller driver
[    7.472709] ehci-platform: EHCI generic platform driver
[    7.482572] ohci_hcd: USB 1.1 'Open' Host Controller (OHCI) Driver
[    7.483386] ohci-platform: OHCI generic platform driver
```

mans0n replied:
pci stuff is not related to usb.
Did you inserted the BT dongle yourself or is it built-in one? If the usb port is not used by on-board module, you can safely disable it as it is not used and it cannot be used.

Lanchon replied:
@mans0n

without both these:
```
&usb3_X {
	status = "okay";
};
```
defs (X = 0 & 1) the USB does not work.

with only &usb3_0..... bootloop!! kernel hangs, and i suppose watchdog gets angry
trying only &usb3_1 now...

> Did you insert the BT dongle yourself

yes i put it there to test if USB responds.

Lanchon replied:
@mans0n ok the usb...

- this works:
```
&usb3_0 {
	status = "okay";
};

&usb3_1 {
	status = "okay";
};
```

- with both defs absent kernel boots but obviously no usb.
- with either def alonge, the kernel hangs and bootloops.

my hypothesis: the clocks are swapped. if only one block is enabled, the kernel clocks the wrong one and talks to the dead, and hell breaks loose.

if true, the dtsi would be wrong here. i'll try swapping the clocks from the dts but i dunno if that overrides the dtsi.

for now seems both blocks must be enabled.

Lanchon replied:
this alone kinda works (no bootloop):
```
&usb3_0 {
	status = "okay";
	clocks = <&gcc USB30_1_MASTER_CLK>;
};
```
but then when i remove the dongle:
```
[  114.207167] usb 1-1: USB disconnect, device number 2
[  124.209588] xhci-hcd xhci-hcd.0.auto: Abort failed to stop command ring: -110
[  124.241641] xhci-hcd xhci-hcd.0.auto: Host halt failed, -110
[  124.241675] xhci-hcd xhci-hcd.0.auto: xHCI host controller not responding, assume dead
[  124.246392] xhci-hcd xhci-hcd.0.auto: HC died; cleaning up
[  124.254392] xhci-hcd xhci-hcd.0.auto: Timeout while waiting for configure endpoint command
```
and stays dead

something is wrong outside of my dts.

dangowrt replied:
can it be that the clocks of the two USB3 host controllers got swapped?
The names kinda indicate that `USB30_1_MASTER_CLK` would belong to `usb3_1` and `USB30_0_MASTER_CLK` to `usb3_0`.

Lanchon replied:
> can it be that the clocks of the two USB3 host controllers got swapped? The names kinda indicate that `USB30_1_MASTER_CLK` would belong to `usb3_1` and `USB30_0_MASTER_CLK` to `usb3_0`.

they seem to be swapped in the dtsi. please read the whole section.

Lanchon replied:
ok...

the block in use is usb3_0 (not _1). however the clocks seem to be swapped somewhere in the included files. adding a clock in the dts does something but does not fully work around the problem: maybe the kernel turns both clocks on and off for each block.

in any case, i'll keep both USB blocks enabled as that seems to be the only effective workaround.

well congrats to us! we just hurt the planet. we wasted amp-hours in building openwrt tens of times, with the prospect of maybe clock off a block and save a few micro-amps in a handful of devices that would never add up to compensate that down the road... no to mention time :)

mans0n replied:
Thanks for the detailed report. I don't understand why the clocks are swapped, but I think it's okay to leave both of the usb nodes turned on for now.
And one thing that just came to mind is that the usb phys (`ss_phy_X` and `hs_phy_X`) might need to be swapped too...or not.

Lanchon replied:
the facts:
1) as the dtsi defines it (no clock added in dts):
   a) if you enable either single block: the kernel hangs on boot, then bootloops.
   b) if you enable both blocks: the USB port seems to work.
2) if you add the "wrong" clock in the dts (in addition to the "right" one in the dts):
   a) if you enable only usb3_0: the USB port works on boot. but as soon as you unplug a USB device, the xhci block stops responding and the driver quickly declares it dead. this might happen as a side effect of usb3_1 also declaring both clocks: the kernel might turn off usb3_1 again, which also causes usb3_0 to die.
   b) if you enable only usb3_1: the USB port work does not work.
   c) if you enable both blocks: the USB port seems to work.

so workable setups seem to be 1b) and 2c). i chose 2c) because:
- no explicit clocks is clearly not working as expected.
- the explicit "wrong" clocks calls attention to this problem. <-- main reason
- somebody that knows more chose this setup:
  - 1b) could have issues that 2c) solves, IDK.
- double clock defs seems not worse than wrong clock defs.

> I think it's okay to leave both of the usb nodes turned on for now
changing the dtsi affects lots of devices so i don't have a choice.

besides, i don't even know their origin: are the dtsi's provided by qualcomm?

Lanchon replied:
@mans0n 
and this is fishy:
```
  hs_phy_0: hs_phy_0 {
   reg = <0x110f8800 0x30>;
  ss_phy_0: ss_phy_0 {
   reg = <0x110f8830 0x30>;
  usb3_0: usb3@110f8800 {
   reg = <0x110f8800 0x8000>;
   dwc3_0: dwc3@11000000 {
    reg = <0x11000000 0xcd00>;

  hs_phy_1: hs_phy_1 {
   reg = <0x100f8800 0x30>;
  ss_phy_1: ss_phy_1 {
   reg = <0x100f8830 0x30>;
  usb3_1: usb3@100f8800 {
   reg = <0x100f8800 0x8000>;
   dwc3_1: dwc3@10000000 {
    reg = <0x10000000 0xcd00>;
```
so:
- all usb 0: 0x110xxxxx
- all usb 1: 0x100xxxxx

weird naming. coupled to the clock issue, smells rotten.

if there is a mixup at all -not silicon errata-, all things mostly static could be implicated:
- clocks
- enables
- resets
- interrupts even

this potential rabbit hole is clearly outside the scope of this port. but do you want to dig deeper? i enabled CCACHE so i'm faring better here, #facepalm. i'd still want to prioritize backporting if possible.

mans0n commented:
And could you add a brief note on how the gmac/switch/ports are interconnected in the commit message?

Lanchon commented:
> > * install instructions: [please comment](https://github.com/openwrt/openwrt/pull/9623#issuecomment-1086827902) on how to proceed.
> 
> It would be the best if you can put all the instructions in the commit message. It seems most of your install script is for checking the stock FW, I think you can extract the essense of it and put the summary in the commit message. And then you can link the detailed (and safer) instructions (which should be copied to the wiki page). Please have a look at [da8428d](https://github.com/openwrt/openwrt/commit/da8428d277cd3373b05330cb3b4f93aef717c5ab).

actually it is mostly all install script: https://github.com/Lanchon/openwrt-tr4400-v2/blob/main/install-recovery.sh

i can forgo making fw_env backups and checks before move (then if moved twice, fw_env is lost forever), but i cannot compress uboot setup that much: it has to at least try to boot from extra/recovery and ubi/kernel, though that is incomplete IMHO. and it does not fit in the required 70/75 columns.

but here is the important thing: the install script requires downloading the uImage to the device, which is not trivial because wget is crippled. so one has to configure sshd first, which is why the install script is not even enough. and if one is downloading uImage, why not download the script too and be done with it?

anyway, that won't work. so i'll just include generic uboot tftp instructions without recovery without backups with the partition move and be done with that.

> precal nvmem-cells is already being used in 22.03 so you don't need to split your patch for it.

ok didnt know that, thanks!

Lanchon commented:
@mans0n @dangowrt, install instructions done!

Lanchon commented on target/linux/ipq806x/patches-5.10/0069-arm-boot-add-dts-files.patch:
```diff
@@ -32,6 +32,7 @@ Signed-off-by: John Crispin <john@phrozen.org>
 +	qcom-ipq8064-wxr-2533dhp.dtb \
 +	qcom-ipq8065-nbg6817.dtb \
 +	qcom-ipq8065-r7800.dtb \
++	qcom-ipq8065-tr4400-v2.dtb \
```
i hand-crafted that diff. i don't know the tooling you use to create the diffs.

mans0n replied:
There's a tool "quilt" for dealing with complex diff, but this diff is simple and I would craft it manually too.

Lanchon replied:
done!

Lanchon commented:
@mans0n @dangowrt 

i think this wraps it up unless further things show up. completely fixing the usb seems to depend on fixings dtsi's which is beyond me. the current workaround works reliably, but i couldn't get reliable operation enabling only one xhci, which might be why the OEM also uses this workaround. i also retested everything (including phy0) just in case.

unfortunately i can't prioritize working on nvmem-cells for caldata right now. since the current approach is correct, i don't see a real problem with that, and other ports are in the same state. i guess i'll come back to that sometime.

thanks!

Lanchon commented:
@mans0n ,

the nvmem-cells commit "seems" to work:
- the athX MACs are set correctly
- the pre-cal data... who know?

thanks!

Lanchon committed:
ipq806x: add support for Arris TR4400 v2 / RAC2V1A

Hardware specs:
  SoC: Qualcomm IPQ8065 (dual core Cortex-A15)
  RAM: 512 MB DDR3
  Flash: 256 MB NAND, 32 MB NOR
  WiFi: QCA9983 2.4 GHz, QCA9984 5 GHz
  Switch: QCA8337
  Ethernet: 5x 10/100/1000 Mbit/s
  USB: 1x USB 3.0 Type-A
  Buttons: WPS, Reset
  Power: 12 VDC, 2.5 A

Ethernet ports:
  1x WAN: connected to eth2
  4x LAN: connected via the switch to eth0 and eth1
          (eth0 is disabled in OEM firmware)

MAC addresses (OEM and OpenWrt):
  fw_env @ 0x00  d4:ab:82:??:??:?a  LAN (eth1)
  fw_env @ 0x06  d4:ab:82:??:??:?b  WAN (eth2)
  fw_env @ 0x0c  d4:ab:82:??:??:?c  WLAN 2.4 GHz (ath1)
  fw_env @ 0x12  d4:ab:82:??:??:?d  WLAN 5 GHz (ath0)
  fw_env @ 0x18  d4:ab:82:??:??:?e  OEM usage unknown (eth0 in OpenWrt)

  OID d4:ab:82 is registered to:
  ARRIS Group, Inc., 6450 Sequence Drive, San Diego CA 92121, US

More info:
  https://openwrt.org/inbox/toh/arris/tr4400_v2

IMPORTANT:

This port requires moving the 'fw_env' partition prior to first boot to
consolidate 70% of the usable space in flash into a contiguous partition.
'fw_env' contains factory-programmed MAC addresses, SSIDs, and passwords.
Its contents must be copied to 'rootfs_1' prior to booting via initramfs.
Note that the stock 'fw_env' partition  will be wiped during sysupgrade.

A writable 'stock_fw_env' partition pointing to the old, stock location
is included in the port to help rolling back this change if desired.

Installation:

- Requires serial access and a TFTP server.
- Fully boot stock, press ENTER, type in:
mtd erase /dev/mtd21
dd if=/dev/mtd22 bs=128K count=1 | mtd write - /dev/mtd21
umount /config && ubidetach -m 23 && mtd erase /dev/mtd23
- Reboot and interrupt U-Boot by pressing a key, type in:
set mtdids 'nand0=nand0'
set mtdparts 'mtdparts=nand0:155M@0x6500000(mtd_ubi)'
set bootcmd 'ubi part mtd_ubi && ubi read 0x44000000 kernel && bootm'
env save
- Setup TFTP server serving initramfs image as 'recovery.bin', type in:
set ipaddr 192.168.1.1
set serverip 192.168.1.2
tftpboot recovery.bin && bootm
- Use sysupgrade to install squashfs image.

This port is based on work done by AmadeusGhost <amadeus@jmu.edu.cn>.

Signed-off-by: Rodrigo Balerdi <lanchon@gmail.com>

Lanchon commented:
@dangowrt your requested change is done. what happens now? should i ask somebody for review or just wait?

Lanchon commented:
this PR is outdated because it does not include the latest USB workaround.

the latest workaround is in the 22.03 branch, and is:
https://github.com/openwrt/openwrt/blob/76e7213547751907e91dbf78456fb45a81f69901/target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-tr4400-v2.dts#L347-L358

i don't want to maintain 2 branches so i'll stay on 22.03. if you can't merge that one, i'll come back here later.

mans0n commented:
AFAICT you should prioritize the master branch. A patch should be backported to stable branch, not the other way around.
Note that I'm not keen to backport stuff other than outright fix.
https://openwrt.org/docs/guide-developer/device-support-policies#backports

Lanchon commented:
yes i had already read that by now, thanks.

this is kind of a fringe case. changes to code that are not fixes are supposed not to be accepted after the code freeze. then again, this is not changing anything at all, it is just adding a new device.

and new devices are generally not backported. but 22.03 is still not out, so this is not really a backport yet.

if i hadn't hit and debugged with @dangowrt the permission issues of the build system for a cpl of weeks (i thought it was my fault port to blame of the time), and if i hadn't spent almost another week with this dtsi issue, this would have been ready a month ago.

oh well... anyway, i posted here because Daniel told me he would consider backporting when i opened the other PR. in the end there is probably not much use in delaying; whenever this is merged, i bet you i will be the sole tester :)

btw, thanks for the thorough review earlier: made me dive and learn into the device tree thing.

Lanchon commented:
@mans0n 
now depends on dtsi fix.
no workarounds of any kind in the device commit.

Lanchon commented:
@mans0n @dangowrt 

hi,

- the changes requested by @dangowrt have long been done.
- the first 2 commits (fix dtsi bugs) were already merged in master by @hauke.
- the 3rd commit (adding the new device) no longer has workarounds for dtsi bugs, has no other review issues pending fix, and is ready for cherrypicking.

thank you all for the help during the porting!

mans0n commented on target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-tr4400-v2.dts:
```diff
@@ -0,0 +1,415 @@
+	};
+};
+
+&ART {
+	compatible = "nvmem-cells";
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	ART_precal_0_ath0_5G: precal@1000 { reg = <0x1000 0x2f20>; };
+	ART_precal_1_ath1_2G: precal@5000 { reg = <0x5000 0x2f20>; };
+};
+
+&fw_env {
+	compatible = "nvmem-cells";
+	#address-cells = <1>;
+	#size-cells = <1>;
+
+	fw_env_macaddr_0_eth1_lan: macaddr@0  { reg = <0x00 0x6>; };
+	fw_env_macaddr_1_eth2_wan: macaddr@6  { reg = <0x06 0x6>; };
+	fw_env_macaddr_2_ath1_2G:  macaddr@c  { reg = <0x0c 0x6>; };
+	fw_env_macaddr_3_ath0_5G:  macaddr@12 { reg = <0x12 0x6>; };
+	fw_env_macaddr_4_unknown:  macaddr@18 { reg = <0x18 0x6>; };
+};
```
One last comment, could you move this into the node definition? And then please remove the partition node labels (`ART:` and `fw_env:`).
https://github.com/openwrt/openwrt/blob/c0a6fefd9b74af03a30810be20f0ce5504447b3e/target/linux/ipq806x/files/arch/arm/boot/dts/qcom-ipq8065-rt4230w-rev6.dts#L208-L231

While at it, could you consider following the convention for the nvmem-cells nodes?
```
precal_ART_1000: precal@1000 {
	reg = <0x1000 0x2f20>;
};

macaddr_fw_env_18: macaddr@18 {
	reg = <0x18 0x6>;
};
```

I would be happy if you add a small note for the MAC addresses in the commit message like 70c12d26ca6e.

Lanchon replied:
@mans0n,
FYI: not much to say about the MACs, except that i haven't seen `fw_env_macaddr_4_unknown` being used by stock fw. stock doesn't set the addresses of interfaces before enabling them, and i've never seen eth0 enabled. i still chose to give it an address over leaving it unconfigured; and since `fw_env_macaddr_4_unknown` is provisioned (the 5 MACs are sequential) and would otherwise go unused, i gave it that.

mans0n replied:
I think it's okay, please just write down about your decision in the commit message.

Lanchon replied:
hi @mans0n 

let me first say that i really appreciate the time you take to review and help with PRs.

unfortunately your review came over a week past my last commit, and by that time i had put the router back together and commissioned it. it's not that i can't grab it back; i can, and i would if there were correctness issues, but i can't justify the time potentially needed -especially if i required serial again- for changes that are style-only.

btw i favor my style, where variables are named according to their usage/content rather than their addresses, negating the need for documenting variable contents in the commit message. i prefer array indices over addresses, and separating concerns so that i don't have to go over the entire partition table to find information that -other than being stored in flash- has not relation to the partition table. but of course style is a matter of opinion entirely.

your criticism while working on this was very much appreciated. it forced me to dive in and learn stuff i had no idea of, and it was fun to do that, which is my real motivation for collaborating in free software. but right now i have very little time and i want to invest it where returns are higher: providing a safer upgrade mechanism for openwrt -as i said i would 2/3 weeks ago. and i haven't even started because of lack of time.

i really don't want to commit untested code, even though they are trivial changes. maybe someone else can apply them. it probably doesn't matter much, as i don't think many people will go though the ordeal of tapping the serial port on this thing.

excuse me for not finding time to reply earlier.

and again, thank you very much for your help.

mans0n replied:
First of all, I also appreciate your efforts and contributions to OpenWrt.

Yeah, I'm sorry for my (even late) nitpicks. I agree that it is rather a cosmetic issue. It would be nice if you can test the change, but it is just a small style-only change after all. I think just a simple build test would suffice in this case.

For the label naming scheme, I sometimes feel that having address in the label make things complicated (e.g. similar devices sharing dtsi but having different addresses), but I don't think using array index is way better in that sense (e.g. what if the location is swapped?). And do we actually need iface name in the label? I would name it something like `macaddr_wlan5g` if I was the first who introduced nvmem-cells to mac address, but these labels were generated by a conversion script and I tend to follow conventions unless there's a good attractive reason. Because if we once accept a non-conventional format then another guy will just copy and paste it and I don't like things getting disorganized. Of course, all these are just my personal opinion.

I'm willing to merge this after the requested change, and I can even amend it myself if you provide me the MAC address list so that I can add it in the commit message. Or you can choose to wait for another core dev to merge this in this state.

Thanks for your patience and endeavor again.

Lanchon replied:
hi @mans0n

i suppose everything is taken care of:

https://github.com/openwrt/openwrt/compare/e31ec46b19d9933e6423e6a443311b67e1300cd4..dd52c04402bed06033f33105c15ee74e054f266e#diff-cb281a4e74086520b1a8bcf5b60c2923ad3988298218315ae442c953b0bf48ba

...plus the commit message

i took the router offline and got a full ubi backup of the deployed configuration and then tested this:
wlan adapters work and all adapters have their correct MACs. (fortunately i didn't need the serial.)

i also updated and tested the 22.03 branch which i hadn't in a while and everything checks out (https://github.com/openwrt/openwrt/pull/9658).

thanks for your help!

mans0n commented:
Thanks! Applied your patch as f8b0010dfb54.

Lanchon commented:
thanl you! and thanks for catching the missing 5.15 stuff :)
